fix(calc): Improve zoom view-jumping

Minion3665 opened the pull request:
Previously, we got significant view jumps in calc by virtue of an incorrect center position. To fix this, I've melded together the old formula (from before commit d6f375c501690370e41ceeec18200599bd33490c) with the new formula. I've gated this to calc as the old formula is calc specific. More testing will be needed to determine if there is any jumping in writer/impress.

This old formula had some pitfalls, as it was made to deal with zooming from the top left of the screen rather than an arbitrary point. Particularly notably, the old formula doesn't deal with anything that is not on the edge of a cell. Therefore, there is still some view jumping, there's just likely to be less-of-it and it'll be a bit more consistent rather than having huge jumps every time on higher positions.

This code still doesn't work in some cases, which I will continue to fix:
- It doesn't work on my iPad, it's unclear why but the symptoms are no change in zoom level. It's possible that tablets use YetAnotherZoomFormula
- It doesn't work on my Mac. The symptoms are being *nearly there* but ever-so-slightly off


Change-Id: I7127d8f0a3156ed9dfb04bdd5fb801c318531269


* Resolves: # <!-- related github issue -->
* Target version: master 

### Summary


### TODO

- [ ] ...

### Checklist

- [ ] I have run `make prettier-write` and formatted the code.
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

Minion3665 committed:
fix(calc): Improve zoom view-jumping

Previously, we got significant view jumps in calc by virtue of an
incorrect center position. To fix this, I've melded together the old
formula (from before commit d6f375c501690370e41ceeec18200599bd33490c)
with the new formula. I've gated this to calc as the old formula is
calc specific. More testing will be needed to determine if there is any
jumping in writer/impress.

There are still minor jumps from the end of the animation to the final
zoom pos, which is particularly noticable in split view, but I think
this may actually be *better* than the animated-to position, so instead
of further iterating on this, I'd like to look and see if we should
change that instead.

This code still doesn't affect tablet view.

Signed-off-by: Skyler Grey <skyler.grey@collabora.com>
Change-Id: I7127d8f0a3156ed9dfb04bdd5fb801c318531269

eszkadev commented on browser/src/layer/tile/CanvasTileLayer.js:
```diff
@@ -489,14 +489,26 @@ L.TileSectionManager = L.Class.extend({
+			newCenter = newTopLeftP.add(paneBounds.getSize().divideBy(2)).multiplyBy(app.dpiScale);
+		} else {
+			const newPaneCenter = new L.Point(
+				(docTopLeft.x - splitPos.x + (paneSize.x + splitPos.x) * 0.5 / scale),
```
`splitPos` exists only in Calc, so no sense to have it in this branch probably? (I understand you don't want to break anything and it's just "old behavior" ?)

Also I don't see anything wrt panes in the Calc branch. Did you test with freezed rows and columns?

eszkadev commented:
would be good to add some cypress test too, existing one for inspiration: https://github.com/CollaboraOnline/online/blob/master/cypress_test/integration_tests/desktop/calc/cell_cursor_spec.js#L63

Minion3665 committed:
refactor: Remove splitPos from non-calc path

In calc, splitPos is used for frozen cells. As those can't exist
in other apps it makes no sense to keep the splitPos terms in the
calculations: they will always be zero

Signed-off-by: Skyler Grey <skyler.grey@collabora.com>
Change-Id: I41b7135408402035773141b4af174e6c3f645ded

Minion3665 committed:
test(zoom): ensure minimal drift from expected

In the previous commit, I changed the zoom algorithm for calc (notably
not writer or impress) to reduce the jumping we get when zooming in/out

Before that commit, we would get some significant jumping. After, we
would get much less.

So, this test is quite evil. I'm going to tell you how it's evil and
help you understand why

1. it's written this way
2. it should be merged despite this evil-ness

Let's start with (1). The test is written this way as a symptom of the
code sucking a bit. The code in the past few commits introduces a
regression where over time zooming in and out adds a small but
compounding error. Basically, it drifts a bit.

The approach to writing this test which would work on non-drifting code
is to have some minimal tolerance which the zoom has to fall within,
and if it does then we could consider it "close enough". That would've
worked before (for some very large tolerance) but as the error here
compounds it won't work.

There are a couple of naïve approaches one could consider for writing
this:
- Nailing down incorrect zoom values
  - This is not ideal, because it means that future commits which make
    this better will start failing the test
- Having a tolerance that increases with more zooms
  - This is better, but by the time you've zoomed in and out a bit
    your tolerance will be so big that it's hard to justify checking...

Instead, I've gone with an approach that calculates the difference it
should see between two zooms and accepts it if the difference is small.
That introduces some complexity with getting past values/etc., but it's
in a test and relatively well-contained...

So, (2) then. That's easy enough. The zoom drift is small enough that -
while obviously noticeable - you feel like you're in about the right
area of the sheet. Before this patch series, even though zooming out
then in would get you to a precise position, it felt like the wrong
precise position. This drift compounds but is otherwise far less
significant than the drift before. I assert that, while the drift
compounding is a bad property for the zoom to have, it is more important
to have a much more minimal drift than to avoid it compounding.

Co-Authored-By: Chris Lord <chris.lord@collabora.com>
Change-Id: Ia575bebb64af49e8ee2c9b91b171aeb00ebcaa5c
Signed-off-by: Skyler Grey <skyler.grey@collabora.com>

Minion3665 commented:
While writing tests for this, I noticed that in its current incarnation it loses a lot of precision sometimes, leading to it being *worse* than the current zoom algorithm at keeping where it's zooming around (the current zoom algorithm will get to ~400 cells out in a single zoom step, but that's the most it'll do and it'll often jump back to stay around the zoom center whereas this one will tend to drift) ... certainly I'm not happy with this

OTOH, it never gets *that far* out in single, regular-sized zoom steps

Minion3665 commented:
Anyway, this leads to the fun effect that it's somewhat difficult to write large-zoom tests that are sensible for it, because they will tend to pass on the previous incarnation if they pass on this one ...

Minion3665 commented:
I rebased this nontrivially and made some additional changes. As is customary, here's the rangediff...

```
1:  c5e02d4d32 ! 1:  e54e50c8ce fix(calc): Improve zoom view-jumping
    @@ Commit message
         Change-Id: I7127d8f0a3156ed9dfb04bdd5fb801c318531269
     
      ## browser/src/layer/tile/CanvasTileLayer.js ##
    -@@ browser/src/layer/tile/CanvasTileLayer.js: L.TileSectionManager = L.Class.extend({
    +@@ browser/src/layer/tile/CanvasTileLayer.js: window.L.TileSectionManager = window.L.Class.extend({
      			return { offset: this._offset, topLeft: docTopLeft };
      		}
      
    --		const newPaneCenter = new L.Point(
    +-		const newPaneCenter = new cool.Point(
     -			(docTopLeft.x - splitPos.x + (paneSize.x + splitPos.x) * 0.5 / scale),
     -			(docTopLeft.y - splitPos.y + (paneSize.y + splitPos.y) * 0.5 / scale));
     +		let newCenter;
    @@ browser/src/layer/tile/CanvasTileLayer.js: L.TileSectionManager = L.Class.extend
     +				.getCorePixelsAtZoom(docTopLeft, zoomScaleAbs);
     +			newCenter = newTopLeftP.multiplyBy(app.dpiScale).add(paneBounds.getSize().subtract(splitPos).divideBy(2));
     +		} else {
    -+			const newPaneCenter = new L.Point(
    ++			const newPaneCenter = new cool.Point(
     +				(docTopLeft.x - splitPos.x + (paneSize.x + splitPos.x) * 0.5 / scale),
     +				(docTopLeft.y - splitPos.y + (paneSize.y + splitPos.y) * 0.5 / scale));
     +			newCenter = this._map.rescale(newPaneCenter, this._map.getZoom(), newZoom);
2:  3863e9df4b ! 2:  c9f012b14a refactor: Remove splitPos from non-calc path
    @@ Commit message
         Change-Id: I41b7135408402035773141b4af174e6c3f645ded
     
      ## browser/src/layer/tile/CanvasTileLayer.js ##
    -@@ browser/src/layer/tile/CanvasTileLayer.js: L.TileSectionManager = L.Class.extend({
    +@@ browser/src/layer/tile/CanvasTileLayer.js: window.L.TileSectionManager = window.L.Class.extend({
      			newCenter = newTopLeftP.multiplyBy(app.dpiScale).add(paneBounds.getSize().subtract(splitPos).divideBy(2));
      		} else {
    - 			const newPaneCenter = new L.Point(
    + 			const newPaneCenter = new cool.Point(
     -				(docTopLeft.x - splitPos.x + (paneSize.x + splitPos.x) * 0.5 / scale),
     -				(docTopLeft.y - splitPos.y + (paneSize.y + splitPos.y) * 0.5 / scale));
     +				(docTopLeft.x - paneSize.x * 0.5 / scale),
3:  1aec250c6c ! 3:  caa3700934 dnm: test(zoom):
    @@ Metadata
     Author: Skyler Grey <skyler.grey@collabora.com>
     
      ## Commit message ##
    -    dnm: test(zoom):
    +    test(zoom): ensure minimal drift from expected
    +
    +    In the previous commit, I changed the zoom algorithm for calc (notably
    +    not writer or impress) to reduce the jumping we get when zooming in/out
    +
    +    Before that commit, we would get some significant jumping. After, we
    +    would get much less.
    +
    +    So, this test is quite evil. I'm going to tell you how it's evil and
    +    help you understand why
    +
    +    1. it's written this way
    +    2. it should be merged despite this evil-ness
    +
    +    Let's start with (1). The test is written this way as a symptom of the
    +    code sucking a bit. The code in the past few commits introduces a
    +    regression where over time zooming in and out adds a small but
    +    compounding error. Basically, it drifts a bit.
    +
    +    The approach to writing this test which would work on non-drifting code
    +    is to have some minimal tolerance which the zoom has to fall within,
    +    and if it does then we could consider it "close enough". That would've
    +    worked before (for some very large tolerance) but as the error here
    +    compounds it won't work.
    +
    +    There are a couple of naïve approaches one could consider for writing
    +    this:
    +    - Nailing down incorrect zoom values
    +      - This is not ideal, because it means that future commits which make
    +        this better will start failing the test
    +    - Having a tolerance that increases with more zooms
    +      - This is better, but by the time you've zoomed in and out a bit
    +        your tolerance will be so big that it's hard to justify checking...
    +
    +    Instead, I've gone with an approach that calculates the difference it
    +    should see between two zooms and accepts it if the difference is small.
    +    That introduces some complexity with getting past values/etc., but it's
    +    in a test and relatively well-contained...
    +
    +    So, (2) then. That's easy enough. The zoom drift is small enough that -
    +    while obviously noticeable - you feel like you're in about the right
    +    area of the sheet. Before this patch series, even though zooming out
    +    then in would get you to a precise position, it felt like the wrong
    +    precise position. This drift compounds but is otherwise far less
    +    significant than the drift before. I assert that, while the drift
    +    compounding is a bad property for the zoom to have, it is more important
    +    to have a much more minimal drift than to avoid it compounding.
     
         Co-Authored-By: Chris Lord <chris.lord@collabora.com>
         Change-Id: Ia575bebb64af49e8ee2c9b91b171aeb00ebcaa5c
    -    DO-NOT-MERGE: This test does not pass, as the zoom feature has some drifting left
    +    Signed-off-by: Skyler Grey <skyler.grey@collabora.com>
    +
    + ## cypress_test/integration_tests/common/desktop_helper.js ##
    +@@ cypress_test/integration_tests/common/desktop_helper.js: function actionOnSelector(name, func) {
    + 	});
    + }
    + 
    ++function cygetScrollbarPosition(type) {
    ++	return cy.cGet('#test-div-' + type + '-scrollbar').then(function($item) {
    ++		cy.wrap(parseInt($item.text()));
    ++	});
    ++}
    ++
    + //type represent horizontal/vertical scrollbar
    + //arr : In both cypress GUI and CLI the scrollposition are slightly different
    + //so we are passing both in array and assert using oneOf
    + function assertScrollbarPosition(type, lowerBound, upperBound) {
    + 	cy.log('>> assertScrollbarPosition - start');
    + 
    +-	cy.cGet('#test-div-' + type + '-scrollbar')
    ++	cygetScrollbarPosition()
    + 		.should(function($item) {
    + 			const x = parseInt($item.text());
    + 			expect(x).to.be.within(lowerBound, upperBound);
     
      ## cypress_test/integration_tests/desktop/calc/zoom_spec.js (new) ##
     @@
    @@ cypress_test/integration_tests/desktop/calc/zoom_spec.js (new)
     +/* global describe it cy beforeEach expect require */
     +
     +var helper = require('../../common/helper');
    -+var calcHelper = require('../../common/calc_helper');
     +var desktopHelper = require('../../common/desktop_helper');
     +
     +const zoomLevels = [
    -+  [20, 760, 540],
    -+  [25, 760, 540],
    -+  [30, 760, 540],
    -+  [35, 760, 540],
    -+  [40, 760, 540],
    -+  [50, 760, 540],
    -+  [60, 760, 540],
    -+  [70, 760, 540],
    -+  [85, 760, 540],
    -+  [100, 760, 540],
    -+  [120, 760, 524],
    -+  [150, 760, 520],
    -+  [170, 760, 520],
    -+  [200, 760, 520],
    -+  [235, 760, 500],
    -+  [280, 740, 500],
    -+  [335, 740, 500],
    -+  [400, 740, 500],
    -+]; // Array<[zoom, expectedHorizontal, expectedVertical]>
    ++  20: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  25: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  30: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  35: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  40: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  50: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  60: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  70: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  85: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  100: { expectedHorizontal: 760, expectedVertical: 540 },
    ++  120: { expectedHorizontal: 760, expectedVertical: 524 },
    ++  150: { expectedHorizontal: 760, expectedVertical: 520 },
    ++  170: { expectedHorizontal: 760, expectedVertical: 520 },
    ++  200: { expectedHorizontal: 760, expectedVertical: 520 },
    ++  235: { expectedHorizontal: 760, expectedVertical: 500 },
    ++  280: { expectedHorizontal: 740, expectedVertical: 500 },
    ++  335: { expectedHorizontal: 740, expectedVertical: 500 },
    ++  400: { expectedHorizontal: 740, expectedVertical: 500 },
    ++]; // Record<zoom, { expectedHortizontal, expectedVertical }>
    ++
    ++const acceptableDrift = 10; // FIXME: there is no "acceptable drift", but there was a change that introduced drift but was otherwise considerably better, so we make this tradeoff. Ideally this would be changed to be a static tolerance
     +
     +describe(['tagdesktop', 'tagnextcloud', 'tagproxy'], 'Test jumping on large cell selectionTest zooming on cells far away from the top left', function() {
     +  beforeEach(function() {
    @@ cypress_test/integration_tests/desktop/calc/zoom_spec.js (new)
     +
     +    cy.log("Zooming in")
     +
    -+    for (const [zoomLevel, horizontal, vertical] of zoomLevels.filter(([level, _, __]) => level > 100)) {
    ++    const zoomEntries = Array.from(zoomLevels.entries());
    ++    let prevZoom = 100;
    ++
    ++    for (const [zoomLevel, { expectedHorizontal, expectedVertical }] of zoomEntries.filter(([level, _]) => level > 100)) {
    ++      const expectedHorizontalDifference = expectedHorizontal - zoomLevels[prevZoom].expectedHorizontal;
    ++      const expectedVerticalDifference = expectedVertical - zoomLevels[prevZoom].expectedVertical;
    ++
     +      cy.log(`Zooming to ${zoomLevel}`);
     +
    -+      desktopHelper.zoomIn();
    ++      desktopHelper.cygetScrollbarPosition('horizontal').then((oldHorizontalPosition) => {
    ++        desktopHelper.cygetScrollbarPosition('vertical').then((oldVerticalPosition) => {
    ++          // This has to be nested because cypress queues operations and then runs them all at once - it has no equivalent of await either
    ++
    ++          desktopHelper.zoomIn();
    ++          desktopHelper.shouldHaveZoomLevel(zoomLevel);
     +
    -+      desktopHelper.shouldHaveZoomLevel(zoomLevel);
    ++          desktopHelper.assertScrollbarPosition('horizontal', oldHorizontalPosition + expectedHorizontalDifference - acceptableDrift, oldHorizontalPosition - expectedHorizontalDifference + acceptableDrift);
    ++          desktopHelper.assertScrollbarPosition('vertical', oldVerticalPosition + expectedVerticalDifference - acceptableDrift, oldVerticalPosition - expectedVerticalDifference + acceptableDrift);
    ++        });
    ++      });
     +
    -+      desktopHelper.assertScrollbarPosition('horizontal', horizontal - 10, horizontal + 10);
    -+      desktopHelper.assertScrollbarPosition('vertical', vertical - 10, vertical + 10);
    ++      prevZoom = zoomLevel;
     +    }
     +
     +    cy.log("Zooming out")
     +
    -+    for (const [zoomLevel, horizontal, vertical] of zoomLevels.toReversed().slice(1)) {
    ++    for (const [zoomLevel, { expectedHorizontal, expectedVertical }] of zoomEntries.toReversed().slice(1)) {
    ++      const expectedHorizontalDifference = expectedHorizontal - zoomLevels[prevZoom].expectedHorizontal;
    ++      const expectedVerticalDifference = expectedVertical - zoomLevels[prevZoom].expectedVertical;
    ++
     +      cy.log(`Zooming to ${zoomLevel}`);
     +
    -+      desktopHelper.zoomOut();
    ++      desktopHelper.cygetScrollbarPosition('horizontal').then((oldHorizontalPosition) => {
    ++        desktopHelper.cygetScrollbarPosition('vertical').then((oldVerticalPosition) => {
    ++          // This has to be nested because cypress queues operations and then runs them all at once - it has no equivalent of await either
    ++
    ++          desktopHelper.zoomIn();
    ++          desktopHelper.shouldHaveZoomLevel(zoomLevel);
     +
    -+      desktopHelper.shouldHaveZoomLevel(zoomLevel);
    ++          desktopHelper.assertScrollbarPosition('horizontal', oldHorizontalPosition + expectedHorizontalDifference - acceptableDrift, oldHorizontalPosition - expectedHorizontalDifference + acceptableDrift);
    ++          desktopHelper.assertScrollbarPosition('vertical', oldVerticalPosition + expectedVerticalDifference - acceptableDrift, oldVerticalPosition - expectedVerticalDifference + acceptableDrift);
    ++        });
    ++      });
     +
    -+      desktopHelper.assertScrollbarPosition('horizontal', horizontal - 10, horizontal + 10);
    -+      desktopHelper.assertScrollbarPosition('vertical', vertical - 10, vertical + 10);
    ++      prevZoom = zoomLevel;
     +    }
     +
     +    cy.log("Zooming in")
     +
     +    for (const [zoomLevel, horizontal, vertical] of zoomLevels.filter(([level, _, __]) => level <= 100).slice(1)) {
    ++      const expectedHorizontalDifference = expectedHorizontal - zoomLevels[prevZoom].expectedHorizontal;
    ++      const expectedVerticalDifference = expectedVertical - zoomLevels[prevZoom].expectedVertical;
    ++
     +      cy.log(`Zooming to ${zoomLevel}`);
     +
    -+      desktopHelper.zoomIn();
    ++      desktopHelper.cygetScrollbarPosition('horizontal').then((oldHorizontalPosition) => {
    ++        desktopHelper.cygetScrollbarPosition('vertical').then((oldVerticalPosition) => {
    ++          // This has to be nested because cypress queues operations and then runs them all at once - it has no equivalent of await either
    ++
    ++          desktopHelper.zoomIn();
    ++          desktopHelper.shouldHaveZoomLevel(zoomLevel);
     +
    -+      desktopHelper.shouldHaveZoomLevel(zoomLevel);
    ++          desktopHelper.assertScrollbarPosition('horizontal', oldHorizontalPosition + expectedHorizontalDifference - acceptableDrift, oldHorizontalPosition - expectedHorizontalDifference + acceptableDrift);
    ++          desktopHelper.assertScrollbarPosition('vertical', oldVerticalPosition + expectedVerticalDifference - acceptableDrift, oldVerticalPosition - expectedVerticalDifference + acceptableDrift);
    ++        });
    ++      });
     +
    -+      desktopHelper.assertScrollbarPosition('horizontal', horizontal - 10, horizontal + 10);
    -+      desktopHelper.assertScrollbarPosition('vertical', vertical - 10, vertical + 10);
    ++      prevZoom = zoomLevel;
     +    }
     +  });
     +});
```

or pictorally by commit, excluding the test commit because that's too big to sensibly include a screenshot here...

<img width="2080" height="934" alt="image" src="https://github.com/user-attachments/assets/9434175a-1f8e-4037-bfdf-5c247898eb72" />

<img width="2084" height="545" alt="image" src="https://github.com/user-attachments/assets/6fcf471a-bd69-404d-9be2-4f0ac6eb45e1" />

eszkadev commented:
@Minion3665 can we close? I guess it's something not finished, out of date already?
